app/src/main/java/com/keylesspalace/tusky/StatusListActivity.kt

```diff
@@ -178,8 +185,9 @@ class StatusListActivity : BottomSheetActivity(), HasAndroidInjector {
             mastodonApi.getFilters().fold(
                 { filters ->
                     mutedFilter = filters.firstOrNull { filter ->
+                        // TODO shouldn't this be an exact match (only one keyword; exactly the hashtag)?
                         filter.context.contains(Filter.Kind.HOME.kind) && filter.keywords.any {
-                            it.keyword == tag
+                            it.keyword == hashedTag
```

Comment:
Ok

(Push pending the 3 api calls or 1 api call discussion)

Reply:
However this case can never be removed (or filters need to be migrated).

Reply:
@Lakoja 

Yes. Thinking about it some more I think filters need to be migrated. Having to special case certain filters deep in the filters code forever is a losing strategy; it's technical debt we'll never be rid of. And I say that as the person who, I think, wrote the code with the bug in the first place.

Here's what I think is a better strategy.

1. Create a new per-account preference that stores a boolean. This boolean is false if the filters haven't been migrated, else true.
2. Check this preference in MainActivity. If it's true, do nothing. If it's false, migrate the filters, and then set the preference to true.

The migration strategy is as follows:

- Fetch and iterate through all the user's filters.
- If the filter title does not start `#` then ignore it. Yes, this will ignore filters if the user changed the title after it was created. If the user does that I assume they're sophisticated enough to know what they're doing.
- Extract the non-`#` part of the filter title (i.e. `#foo` becomes `foo`). If a keyword in the filter matches that then prepend the ''#', and save the filter.

We can mention this in the release notes and link to a longer blog post about it.

Now, the mute/unmute code can assume that filtered tags always have a `#` and can be simpler.

---

Optional bonus extra work: In the filter editing code, put a check in if the user tries to edit any filter where the title starts `#`. The dialog can explain that these filters are created to mute hashtags, and that editing them may cause problems.

Or, disallow editing these filters entirely -- separate them visually, and if the user tries to edit it (e.g., taps on it) put up a dialog that says this is a filter that mutes a hashtag, and the only operation that can be done is to delete it.

Let me know if you would prefer me to do that work. This is the last thing that's blocking the 24.0 release, so I'd like to get this resolved soon.

Reply:
That last idea ("disallow editing these filters entirely") might fit in which how followed hashtags are displayed.

The user can go to "Account preferences" and there's a "Followed hashtags" section.

Maybe there should be a "Muted hashtags" section as well, which shows just the filters that affect hashtags. As before, the only options would be to unmute the hashtag (so, delete the filter), or change the filter type from warn to hide, and vice-versa.

I'm not completely sold on this idea -- if the user manages their filters from apps other than Tusky there's the possibility to get things in to a weird state that'll be difficult to recover from within Tusky.

But it might also be easier for users to understand.

Reply:
How about this (just commited):

Only compare the title.

This has the advantages:
- Works with old and new code
- Thus does not require migration (which is _quite_ complex as it involves the api)
- Possibility of some failure (someone editing that filter) is rather limited
- Leaves the possibility to still do some migration/fix in the future (and/or limit editing of those filters)

What is the state of "V1" here? Where is that needed or used?
(Because this fix does not work for that and also cannot be migrated?)